### ISSUE TITLE
Renaming NetworkRequest protocol?

### ISSUE THREAD

**Issue by tylermilner** [BottleRocketStudios/iOS-Hyperspace#36]
I've been thinking about this for a while and wanted to open up discussion. I've personally never really been a fan of the `Any...<T>` convention used for type-erased structs. Instead of `AnyNetworkRequest<T>`, I think it would be nicer to be able to just write `NetworkRequest<T>`. To accomplish this, we would need to rename the `NetworkRequest` protocol to something else. Some ideas I have are:
* `RequestRepresentable`
* `NetworkRequestRepresentable`
* `NetworkRequestProtocol`

This would of course be a breaking change so it'd likely go into a `2.0.0` release. See the examples below for a peak of how this might look.

**Defining something that conforms to `RequestRepresentable`:**
```swift
struct CreatePostRequest: RequestRepresentable {
    // Define the model we want to get back
    typealias ResponseType = Post
    typealias ErrorType = AnyError
    
    // Define NetworkRequest property values
    var method: HTTP.Method = .post
    var url = URL(string: "https://jsonplaceholder.typicode.com/posts")!
    var queryParameters: [URLQueryItem]?
    var headers: [HTTP.HeaderKey: HTTP.HeaderValue]? = [.contentType: .applicationJSON]
    var body: Data? {
        let encoder = JSONEncoder()
        return try? encoder.encode(newPost)
    }
    
    // Define any custom properties needed
    private let newPost: NewPost
    
    // Initializer
    init(newPost: NewPost) {
        self.newPost = newPost
    }
}
```

**Using `NetworkRequest<T>`:**
```swift
let createPostRequest = NetworkRequest<Post>(method: .post,
                                             url: URL(string: "http://jsonplaceholder.typicode.com/posts")!,
                                             headers: [.contentType: .applicationJSON],
                                             body: postBody)
```

**Comment by wmcginty** [comment]
I like the idea a lot, my only reservation is moving away from the example of the standard library with [AnyHashable](https://developer.apple.com/documentation/swift/anyhashable).

I can be convinced either way.

**Comment by tylermilner** [comment]
Yeah, I've thought about that too. Apple has kind of made "Any" the standard here. Maybe even shortening to `AnyRequest<T>` instead of `AnyNetworkRequest<T>` (change `NetworkRequest` protocol to just `Request` as well) would be somewhat of an improvement.

**Comment by wmcginty** [comment]
Is this something we want to try and include for 2.0?

**Comment by tylermilner** [comment]
I think so. I'd like more people to weigh-in with their thoughts though.

**Comment by earlgaspard** [comment]
I could get onboard with dropping the "Any".

**Comment by tylermilner** [comment]
So far I'm leaning towards:
* `NetworkRequest` protocol --> `RequestRepresentable`
* `AnyNetworkRequest<T>` struct --> `Request<T>`

Anyone else have any opinions? @BottleRocketStudios/ios-developers

**Comment by GrandLarseny** [comment]
To provide a bit of pushback, I believe the intended purpose is to differentiate naming a specific, concrete implementation of a protocol from the general idea of the protocol. I do agree that seeing `Any` as a prefix all over the place is ugly, so maybe there's some other way to convey the desired semantic meaning without resorting to a prefix.

It seems like above we're moving towards calling out the protocols instead of the concrete implementations. Could make a lot of sense.

**Comment by wmcginty** [comment]
I'm going to play devil's advocate here and suggest the following (taking the [API naming guidelines](https://swift.org/documentation/api-design-guidelines/#naming) into account strictly):

`NetworkRequest` ->` Request`
`AnyNetworkRequest<T>` -> `AnyRequest<T>`

**Comment by WSTurner** [comment]
I like the idea of:
`NetworkRequest` ->`Request`
`AnyNetworkRequest<T>` ->`AnyRequest<T>`

Apple seems pretty consistent with their naming of type-erased values and I think retaining that consistency is a good idea.

**Comment by wmcginty** [comment]
Under review in #52

**Comment by wmcginty** [comment]
Merged in #52
